src/apphosting/yaml.ts

```diff
@@ -0,0 +1,133 @@
+import { basename, dirname } from "path";
+import { readFileFromDirectory, wrappedSafeLoad } from "../utils";
+import { Config, Env, store } from "./config";
+import * as yaml from "yaml";
+import * as jsYaml from "js-yaml";
+import { fileExistsSync } from "../fsutils";
+
+export type EnvironmentVariable = Omit<Env, "secret">;
+export type Secret = Omit<Env, "availability" | "value">;
+
+/**
+ * AppHostingYamlConfig is an object representing an Apphosting Yaml configuration
+ * present in the user's codebase (i.e 'apphosting.yaml', 'apphosting.staging.yaml', etc).
+ *
+ * This class is used to abstract away the parsing and handling logic of these
+ * yaml files.
+ */
+export class AppHostingYamlConfig {
+  filePath?: string;
+  _loadedAppHostingYaml: Config;
+  private _environmentVariables: Map<string, EnvironmentVariable>;
+  private _secrets: Map<string, Secret>;
+
+  constructor(filePath?: string) {
+    this.filePath = filePath;
+    this._loadedAppHostingYaml = {};
+    this._environmentVariables = new Map();
+    this._secrets = new Map();
+  }
+
+  async init() {
```

Comment:
optional but I like to avoid introducing state into my data objects. Not knowing whether an object is fully "initialized" or not leads to awkward code like the helper later in this file.

The danger here is someone new to the codebase uses this constructor without calling `init` after, leading to all sorts of fun behavior.

How about a factory method?

export class ... {
  // Loads a config from file
  static loadFromFile(filePath): Promise<AppHostingYamlConfig> {..}
  
  // Creates an empty config
  static empty(): AppHostingYamlConfig {..}
 
  /* use one of the factory methods above */
  private constructor(..) {..}
}

Reply:
totally makes sense, refactored it to use the Factory pattern